File: src/VisualBasic/CodeCracker/Usage/DisposableFieldNotDisposedCodeFixProvider.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeActions
Imports Microsoft.CodeAnalysis.CodeFixes
Imports Microsoft.CodeAnalysis.Formatting
Imports Microsoft.CodeAnalysis.Simplification
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Microsoft.CodeAnalysis.VisualBasic.VisualBasicExtensions

Namespace Usage
    <ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=NameOf(DisposableFieldNotDisposedCodeFixProvider)), Composition.Shared>
    Public Class DisposableFieldNotDisposedCodeFixProvider
        Inherits CodeFixProvider

        Public Overrides Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
            Dim root = Await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(False)
            Dim diagnostic = context.Diagnostics.First
            Dim span = diagnostic.Location.SourceSpan
            Dim variableDeclarator = root.FindToken(span.Start).Parent.FirstAncestorOrSelf(Of VariableDeclaratorSyntax)()
            context.RegisterCodeFix(CodeAction.Create("Dispose field '" & variableDeclarator.Names.First().ToString(),
                                              Function(c) DisposeField(context.Document, variableDeclarator, c)),
                            diagnostic)

        End Function

        Public Overrides NotOverridable ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String) = ImmutableArray.Create(DiagnosticId.DisposableFieldNotDisposed_Created.ToDiagnosticId(), DiagnosticId.DisposableFieldNotDisposed_Returned.ToDiagnosticId())

        Public Overrides Function GetFixAllProvider() As FixAllProvider
            Return WellKnownFixAllProviders.BatchFixer
        End Function

        Private Async Function DisposeField(document As Document, variableDeclarator As VariableDeclaratorSyntax, cancellationToken As CancellationToken) As Task(Of Document)
            Dim semanticModel = Await document.GetSemanticModelAsync(cancellationToken)
            Dim type = variableDeclarator.FirstAncestorOrSelf(Of ClassBlockSyntax)
            Dim typeSymbol = semanticModel.GetDeclaredSymbol(type)
            Dim newTypeImplementingIDisposable = AddIDisposableImplementationToType(type, typeSymbol)
            Dim newTypeWithDisposeMethod = AddDisposeDeclarationToDisposeMethod(variableDeclarator, newTypeImplementingIDisposable, typeSymbol)
            Dim root = Await document.GetSyntaxRootAsync()
            Dim newRoot = root.ReplaceNode(type, newTypeWithDisposeMethod)
            Dim newDocument = document.WithSyntaxRoot(newRoot)
            Return newDocument
        End Function

        Private Shared Function AddIDisposableImplementationToType(type As ClassBlockSyntax, typeSymbol As INamedTypeSymbol) As ClassBlockSyntax
            Dim iDisposableInterface = typeSymbol.AllInterfaces.FirstOrDefault(Function(i) i.ToString.EndsWith("IDisposable"))
            If iDisposableInterface IsNot Nothing Then Return type
            Dim implementIdisposable = SyntaxFactory.ImplementsStatement(SyntaxFactory.ParseName("System.IDisposable")).
                WithTrailingTrivia(SyntaxFactory.CarriageReturnLineFeed).
                WithAdditionalAnnotations(Simplifier.Annotation)

            Dim newImplementsList = If(Not type.Implements.Any(),
            type.Implements.Add(implementIdisposable),
            New SyntaxList(Of ImplementsStatementSyntax)().
                              Add(implementIdisposable))

            Dim newType = type.WithImplements(newImplementsList).
            WithAdditionalAnnotations(Formatter.Annotation)
            Return newType
        End Function

        Private Shared Function AddDisposeDeclarationToDisposeMethod(variableDeclarator As VariableDeclaratorSyntax, type As ClassBlockSyntax, typeSymbol As INamedTypeSymbol) As TypeBlockSyntax
            Dim disposableMethod = typeSymbol.GetMembers("Dispose").OfType(Of IMethodSymbol).FirstOrDefault(Function(d) d.Arity = 0)
            Dim disposeStatment = SyntaxFactory.ParseExecutableStatement(variableDeclarator.Names.First().ToString() & ".Dispose()").
                WithTrailingTrivia(SyntaxFactory.CarriageReturnLineFeed)

            Dim newType As TypeBlockSyntax
            If disposableMethod Is Nothing Then
                Dim disposeMethod = SyntaxFactory.SubBlock(SyntaxFactory.SubStatement("Dispose").
                    WithParameterList(SyntaxFactory.ParameterList()).
                    WithImplementsClause(SyntaxFactory.ImplementsClause(SyntaxFactory.QualifiedName(SyntaxFactory.IdentifierName("IDisposable"), SyntaxFactory.IdentifierName("Dispose")))).
                    WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword))),
                    New SyntaxList(Of StatementSyntax)().Add(disposeStatment)).
                    WithAdditionalAnnotations(Formatter.Annotation)

                newType = type.AddMembers(disposeMethod)
            Else
                Dim existingDisposeMethod = TryCast(disposableMethod.DeclaringSyntaxReferences.FirstOrDefault?.GetSyntax().Parent, MethodBlockSyntax)
                If type.Members.Contains(existingDisposeMethod) Then
                    Dim newDisposeMethod = existingDisposeMethod.AddStatements(disposeStatment).
                    WithAdditionalAnnotations(Formatter.Annotation)

                    ' Ensure the dispose method includes the implements clause
                    Dim disposeStatement = newDisposeMethod.SubOrFunctionStatement
                    Dim disposeStatementTrailingTrivia = disposeStatement.GetTrailingTrivia()
                    If disposeStatement.ImplementsClause Is Nothing Then
                        disposeStatement = disposeStatement.
                        WithoutTrailingTrivia().
                        WithImplementsClause(SyntaxFactory.ImplementsClause(SyntaxFactory.QualifiedName(SyntaxFactory.IdentifierName("IDisposable"), SyntaxFactory.IdentifierName("Dispose")))).
                        NormalizeWhitespace(" ").
                        WithTrailingTrivia(disposeStatementTrailingTrivia).
                        WithAdditionalAnnotations(Formatter.Annotation)

                        newDisposeMethod = newDisposeMethod.ReplaceNode(newDisposeMethod.SubOrFunctionStatement, disposeStatement)
                    End If
                    newType = type.ReplaceNode(existingDisposeMethod, newDisposeMethod)
                Else
                    Dim fieldDeclaration = variableDeclarator.Parent
                    Dim newFieldDeclaration = fieldDeclaration.
                        WithTrailingTrivia(SyntaxFactory.CommentTrivia("' Add " & disposeStatment.ToString() & " to the Dispose method on the partial file." & vbCrLf))

                    newType = type.ReplaceNode(fieldDeclaration, newFieldDeclaration)
                End If
            End If
            Return newType
        End Function
    End Class
End Namespace

File: src/CSharp/CodeCracker/Usage/DisposableFieldNotDisposedCodeFixProvider.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Extensions;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Formatting;
using Microsoft.CodeAnalysis.Simplification;
using System.Collections.Immutable;
using System.Composition;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace CodeCracker.CSharp.Usage
{
    [ExportCodeFixProvider(nameof(DisposableFieldNotDisposedCodeFixProvider), LanguageNames.CSharp), Shared]
    public class DisposableFieldNotDisposedCodeFixProvider : CodeFixProvider
    {
        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
            ImmutableArray.Create(DiagnosticId.DisposableFieldNotDisposed_Created.ToDiagnosticId(), DiagnosticId.DisposableFieldNotDisposed_Returned.ToDiagnosticId());

        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;

        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
            var diagnostic = context.Diagnostics.First();
            var diagnosticSpan = diagnostic.Location.SourceSpan;
            var variableDeclarators = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<VariableDeclaratorSyntax>();
            foreach (var variableDeclarator in variableDeclarators)
                context.RegisterCodeFix(CodeAction.Create($"Dispose field '{variableDeclarator.Identifier.Value}'", c => DisposeFieldAsync(context.Document, variableDeclarator, c)), diagnostic);
        }

        private async Task<Document> DisposeFieldAsync(Document document, VariableDeclaratorSyntax variableDeclarator, CancellationToken cancellationToken)
        {
            var semanticModel = await document.GetSemanticModelAsync(cancellationToken);
            var type = variableDeclarator.FirstAncestorOrSelf<TypeDeclarationSyntax>();
            var typeSymbol = semanticModel.GetDeclaredSymbol(type);
            var newTypeImplementingIDisposable = AddIDisposableImplementationToType(type, typeSymbol);
            var newTypeWithDisposeMethod = AddDisposeDeclarationToDisposeMethod(variableDeclarator, newTypeImplementingIDisposable, typeSymbol);
            var root = await document.GetSyntaxRootAsync();
            var newRoot = root.ReplaceNode(type, newTypeWithDisposeMethod);
            var newDocument = document.WithSyntaxRoot(newRoot);
            return newDocument;
        }

        private static TypeDeclarationSyntax AddIDisposableImplementationToType(TypeDeclarationSyntax type, INamedTypeSymbol typeSymbol)
        {
            var iDisposableInterface = typeSymbol.AllInterfaces.FirstOrDefault(i => i.ToString() == "System.IDisposable");
            if (iDisposableInterface != null) return type;
            var newBaseList = type.BaseList != null
                ? type.BaseList.AddTypes(SyntaxFactory.SimpleBaseType(SyntaxFactory.ParseName("System.IDisposable").WithAdditionalAnnotations(Simplifier.Annotation)))
                : SyntaxFactory.BaseList(SyntaxFactory.SeparatedList(new BaseTypeSyntax[] {
                        SyntaxFactory.SimpleBaseType(SyntaxFactory.ParseName("System.IDisposable").WithAdditionalAnnotations(Simplifier.Annotation)) }));
            TypeDeclarationSyntax newType = ((dynamic)type)
                .WithBaseList(newBaseList)
                .WithIdentifier(SyntaxFactory.Identifier(type.Identifier.Text));//this line is stupid, it is here only to remove the line break at the end of the identifier that roslyn for some reason puts there
            newType = newType.WithAdditionalAnnotations(Formatter.Annotation);//can't chain because this would be an ext.method on a dynamic type
            return newType;
        }

        private static TypeDeclarationSyntax AddDisposeDeclarationToDisposeMethod(VariableDeclaratorSyntax variableDeclarator, TypeDeclarationSyntax type, INamedTypeSymbol typeSymbol)
        {
            var disposableMethod = typeSymbol.GetMembers("Dispose").OfType<IMethodSymbol>().FirstOrDefault(d => d.Arity == 0);
            var disposeStatement = SyntaxFactory.ParseStatement($"{variableDeclarator.Identifier.ToString()}.Dispose();");
            TypeDeclarationSyntax newType;
            if (disposableMethod == null)
            {
                var disposeMethod = SyntaxFactory.MethodDeclaration(SyntaxFactory.PredefinedType(SyntaxFactory.Token(SyntaxKind.VoidKeyword)), "Dispose")
                      .WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.PublicKeyword)))
                      .WithBody(SyntaxFactory.Block(disposeStatement))
                      .WithAdditionalAnnotations(Formatter.Annotation);
                newType = ((dynamic)type).AddMembers(disposeMethod);
            }
            else
            {
                var existingDisposeMethod = (MethodDeclarationSyntax)disposableMethod.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax();
                if (type.Members.Contains(existingDisposeMethod))
                {
                    var newDisposeMethod = existingDisposeMethod.AddBodyStatements(disposeStatement)
                        .WithAdditionalAnnotations(Formatter.Annotation);
                    newType = type.ReplaceNode(existingDisposeMethod, newDisposeMethod);
                }
                else
                {
                    //we will simply anotate the code for now, but ideally we would change another document
                    //for this to work we have to be able to fix more than one doc
                    var fieldDeclaration = variableDeclarator.Parent.Parent;
                    var newFieldDeclaration = fieldDeclaration.WithTrailingTrivia(SyntaxFactory.ParseTrailingTrivia($"//add {disposeStatement.ToString()} to the Dispose method on another file.").AddRange(fieldDeclaration.GetTrailingTrivia()))
                        .WithLeadingTrivia(fieldDeclaration.GetLeadingTrivia());
                    newType = type.ReplaceNode(fieldDeclaration, newFieldDeclaration);
                }
            }
            return newType;
        }
    }
}